File: ex/hexagonal/src/test/kotlin/com/ktb/shop/application/ListProductsByCategoryTest.kt
package com.ktb.shop.application

import com.ktb.shop.domain.CategoryId
import com.ktb.shop.domain.Product
import com.ktb.shop.domain.ProductId
import com.ktb.shop.domain.port.ProductCatalog
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class ListProductsByCategoryTest {

    @Test
    fun `returns the products in a category`() {
        val mug = Product(ProductId("p1"), CategoryId("c1"), "Mug", 19900)
        val catalog = object : ProductCatalog {
            override fun byCategory(category: CategoryId) =
                if (category == CategoryId("c1")) listOf(mug) else emptyList()
            override fun byId(id: ProductId): Product? = null
            override fun all() = listOf(mug)
        }

        assertEquals(listOf(mug), ListProductsByCategory(catalog).handle("c1"))
    }
}

File: ex/hexagonal-go/internal/application/list_products_by_category_test.go
package application

import (
	"context"
	"testing"

	"github.com/ktb/hexagonal-shop/internal/domain"
)

type fixedCatalog struct {
	products []domain.Product
}

func (c *fixedCatalog) ByCategory(_ context.Context, category domain.CategoryID) ([]domain.Product, error) {
	var out []domain.Product
	for _, p := range c.products {
		if p.CategoryID == category {
			out = append(out, p)
		}
	}
	return out, nil
}

func (c *fixedCatalog) ByID(_ context.Context, id domain.ProductID) (*domain.Product, error) {
	for i := range c.products {
		if c.products[i].ID == id {
			return &c.products[i], nil
		}
	}
	return nil, nil
}

func (c *fixedCatalog) All(_ context.Context) ([]domain.Product, error) {
	return c.products, nil
}

func TestReturnsTheProductsInACategory(t *testing.T) {
	mug, err := domain.NewProduct(domain.NewProductID("p1"), domain.NewCategoryID("c1"), "Mug", 19900)
	if err != nil {
		t.Fatal(err)
	}
	uc := NewListProductsByCategory(&fixedCatalog{products: []domain.Product{mug}})

	got, err := uc.Handle(context.Background(), "c1")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(got) != 1 || got[0] != mug {
		t.Fatalf("expected [mug], got %v", got)
	}

	empty, err := uc.Handle(context.Background(), "c-empty")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(empty) != 0 {
		t.Fatalf("expected empty list, got %v", empty)
	}
}
